src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -418,6 +422,10 @@ async ValueTask<int> Core(Memory<byte> buffer, CancellationToken cancellationTok
                             int n = await _stream.ReadAsync(new Memory<byte>(_buffer, 0, _buffer.Length), cancellationToken).ConfigureAwait(false);
                             if (n <= 0)
                             {
+                                if (!_inflater.Finished() && _inflater.NonEmptyInput() && _inflater.AvailableOutput > 0)
```

Comment:
I understand the condition as "if we haven't finished inflating AND there's partial input set AND there's already output available; we are in the bad state and we should throw". 
Is it possible (i.e: an scenario) where there could be partial input but not available output?

Reply:
Can you please double check test-code-coverage of the three conditions. If I remove `_inflater.AvailableOutput > 0` all tests still pass.

Reply:
Yes, without `_inflater.AvailableOutput > 0`, the following test fails.

```
  System.IO.InvalidDataException : Found invalid data while decoding.

  Stack Trace: 
DeflateStream.ThrowGenericInvalidData() line 348
DeflateStream.ReadCore(Span`1 buffer) line 280
DeflateStream.Read(Byte[] buffer, Int32 offset, Int32 count) line 235
ZipFileTestBase.ReadAllBytes(Stream stream, Byte[] buffer, Int32 offset, Int32 count) line 83
ZipFileTestBase.StreamsEqual(Stream ast, Stream bst, Int32 blocksToRead) line 142
ZipFileTestBase.StreamsEqual(Stream ast, Stream bst) line 111
ZipFile_Create.CreateFromDirectory_IncludeBaseDirectory() line 42
```

This is the reason I added an async version of the test in this PR, to cover the async code path.